### ISSUE TITLE
Using images with with nodata pixels

### ISSUE THREAD

**Issue by GinaONeil** [nshaud/DeepNetsForEO#35]
Hello, I am using the pytorch version of your code to detect wetlands from a 3-band geotiff made up of topographic indices. My verification data is made up of manually surveyed wetland delineations, so it has an irregular shape. I use the wetland surveys as ground truth wetland and nonwetland areas for training and testing purposes by splitting the surveyed area into separate tiles and randomly splitting these into the two groups. Of course, some of the training tiles end up having no data pixels, representing land cover that is not known to be either wetland or nonwetland because it extends further than the limits of the original wetland surveys. If I limit my training set to only those tiles without nodata pixels, I get reasonable predictions. However, I'm already VERY limited in my training data size, relative to most deep learning applications, and would rather include all training data possible by letting the code just ignore these sets of pixels. 

So far I've tried:
1. setting the nodata pixels to a value outside of the range(len(LABELS)) i.e., nonwetlands = 0, wetlands = 1 and nodata >1 or < 0
2. including nodata value in the list of target labels i.e., nodata=0, nonwetlands=1, wetlands=2 and then adjusting the WEIGHTS parameter to give nodata a weight of 0.
3. Both 1 and 2 PLUS setting ignore_index = [nodata value] when calling cross_entropy()

These attempts either cause loss = nan or produce poor results after a couple of epochs (predict everything as nonwetland)

Do you have other ideas for signaling to pytorch to ignore a certain type of pixel? I'm very new to deep learning and pytorch so any insight would be very helpful! Thanks! Example training image and corresponding label tile that have nodata pixels

![training_EX](https://user-images.githubusercontent.com/17127263/54716679-a4f9f680-4b2c-11e9-8db2-6cdc7ddafef4.jpg)
![image_EX](https://user-images.githubusercontent.com/17127263/54716683-a75c5080-4b2c-11e9-940c-adfd175d6940.jpg)

**Comment by nshaud** [comment]
Ah, this is a very good question actually. Of course, it makes sense to set the ground truth value for NODATA pixels to a index that will be ignored during training. However, the problem is that if you feed images with NODATA pixels to the network, its field of view will encompass those pixels. I don't know how your NODATA values are encoded, but if they are NaNs they will propagate everywhere, up to the loss. This is something that came up when we tried to work with Sentinel-2 data.

One solution is to replace all NODATA pixels by zeroes in the image and using what you already did so that the pixels are masked when computing the loss. This will effectively act like zero-padding.

Another solution is to change how we sample patches from the images. We can provide a mask argument to only extract patches that cover the surveyed area and that do no contain NODATA pixels.

A combination of both is implemented in the SegNet_S2_Philab.ipynb notebook (see `get_random_pos` function).

Best of luck.

**Comment by GinaONeil** [comment]
Thank you! I'll try implementing this and update here

**Comment by GinaONeil** [comment]
Thanks for your suggestion, this seems to have worked! I have not yet implemented changes to the code that will restrict random window sampling to areas with at least 80% real data pixels, as done in SegNet_S2Philab.ipynb (please correct me if I misinterpreted). But by replacing NaN values in both my input data and ground truth images with 0s, and by setting the weight for the '0' target class to be 0, my predictions have improved! Thanks again.

**Comment by nshaud** [comment]
Glad I could help. I'm closing this question for now, feel free to update it if you implement the restriction on the window sampling.
